src/charm.py

```diff
@@ -0,0 +1,477 @@
+            event.fail("Unable to add user.")
+            return
+        event.set_results({"initial-password": password})
+
+    def _reset_services_or_defer(self, event: ops.SecretChangedEvent):
+        try:
+            self._replan(self.unit.get_container(self.background_service_name))
+        except ops.pebble.ConnectionError:
+            self.unit.status = ops.WaitingStatus(
+                f"Unable to connect to {self.background_service_name}"
+            )
+            event.defer()
+        except ops.pebble.ChangeError as e:
+            self.unit.status = ops.BlockedStatus(
+                f"Could not start {self.background_service_name}: {e}"
+            )
+            event.defer()
+        try:
+            self._replan(self.unit.get_container(self.web_service_name))
+        except ops.pebble.ConnectionError:
+            self.unit.status = ops.WaitingStatus(f"Unable to connect to {self.web_service_name}")
+            event.defer()
+        except ops.pebble.ChangeError as e:
+            self.unit.status = ops.BlockedStatus(f"Could not start {self.web_service_name}: {e}")
+            event.defer()
+
+    def _on_pebble_ready(self, event: ops.PebbleReadyEvent):
+        """Handle pebble-ready event."""
+        container = event.workload
+        try:
+            self._replan(container)
+        except ops.pebble.ConnectionError:
+            self.unit.status = ops.WaitingStatus(f"Unable to connect to {container.name}")
+            event.defer()
+        except ops.pebble.ChangeError as e:
+            self.unit.status = ops.BlockedStatus(f"Could not start {container.name}: {e}")
+            event.defer()
+
+    def _replan(self, container: ops.Container):
+        summary, command, update_version = self.container_details[container.name]
+        logger.debug("Updating plan for %s (%s -> %s)", container.name, summary, command)
+        layer = ops.pebble.Layer(
+            {
+                "summary": summary,
+                "description": "Service for Takahē ActivityPub instance",
+                "services": {
+                    container.name: {
+                        "override": "replace",
+                        "summary": summary,
+                        "command": command,
+                        "startup": "enabled",
+                        "environment": self._takahē_env,
+                    },
+                },
+            }
+        )
+        new_layer = layer.to_dict()
+        services = container.get_plan().to_dict().get("services", {})
+        if services != new_layer.get("services"):
+            container.add_layer(container.name, layer, combine=True)
+            logger.info("Added updated layer %r to Pebble plan.", container.name)
+            container.replan()
+            logger.info("Updated services.")
+
+        if update_version:
+            self.unit.set_workload_version(self.workload_version)
+
+    @property
+    def workload_version(self):
```

Comment:
> Personally I think of properties as basically free, things that probably can't raise exceptions, partly because it's unclear that just accessing a property will have a cost and raise exceptions. Thoughts on making these type of properties methods instead, like `get_workload_version`?

Good call. I generally prefer to only have properties like that too. I was lead down this track by the tutorial - I wasn't basing the charm on it like I did my first one, but there were still some patterns that I lifted, like this one. I agree this would be better as a method (likewise in the tutorial, but again, not at the cost of ~11 PRs...).

Reply:
Hmm, yeah -- if we start helping maintain that tutorial we'll definitely have to figure out how to make changes less painful (otherwise they won't happen, and fair enough).